Flow event subclasses by sharon.

sharonyang opened the pull request:
Some TODOs are left for TCP algorithm. Need to verify some logics. Please comment!

sharonyang committed:
Initial flow event subclasses design. Not completed.

sharonyang committed:
Touch up to event subclasses. Ready for PR.

Finish implement most flow event subclasses. Blocked on TCP to
finalized functions for ReceiveEvent.

sharonyang committed:
Added minor comment changes to pass PyCharm's nit comment check.

Also changed import for packet.py to relative import.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        self.packet_ids = packet_ids
+        self.packets_to_send = []
+
+    def create_data_packet(self, packet_id, curr_time):
```
Seems like this function is only used once so maybe just inline the DataPacket creation.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
-# TODO(team): Flow subclasses for at least two TCP algorithms.
+
+
+class InitiateFlowEvent(Event):
```
This is called during the initializer code, and that code does not know what "packet_ids" is.

What needs to happen is that, during run(), you need to figure out the first set of packets to send for the Flow. These should be {0, ..., N} where N is determined by the window size of the flow (defaults to 1 during construction) and the number of packets left to send.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param Statistics statistics: the Statistics to update.
+        """
+        for curr_packet in self.packet_ids:
+            if not self.flow.packet_id_exceeds_data(curr_packet):
```
You also can't exceed the window size. Maybe this is what packet_ids is doing, but that should be something decided by the flow and not passed into the Event as a constructor param.

sharonyang replied:
Should be addressed with the new set up for packet_ids.
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param float curr_time: simulation time of this function call.
+        :return DataPacket
+        """
+        packet = DataPacket(packet_id, self.flow.flow_id,
```
Wouldn't a constructor call require specifying the param names since these are "optional" in the constructor? E.g. packet_id=packet_id. I forget how Python does this.

sharonyang replied:
So without default values, the caller would get an error if those parameters are not specified, which was what I was trying to do here.
But this function is removed anyway, so done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param MainEventLoop main_event_loop: main loop.
+        :param Statistics statistics: the Statistics to update.
+        """
+        if len(set(self.flow.packets_in_transit) \
```
You will have to do set(...).union(set(...)). Python doesn't recognize + with sets means set addition.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        if len(set(self.flow.packets_in_transit) \
+            + set(self.flow.packets_to_send)) \
+            > self.flow.window_size_packets:
+                # TODO(team): What do we want to throw here?
```
raise a ValueError or Exception for now so we know about this when it happens.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
```
Just schedule another FlowTimeoutPacketEvent, don't renew by going through the Queue. Events shouldn't be able to do anything other than add to the Event queue.

We will deal with this by making the timeout event do nothing if the packet's ACK has been received (packet not in transit).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+            device_link_event = DeviceToLinkEvent(curr_packet, False)
+            main_event_loop.schedule_event_with_delay(device_link_event, 0.0)
+
+        # Go through the queue and remove any previously scheduled
```
Just schedule another timeout event. If we get to it and there's no need to do anything, then we'll just skip over it.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+            timeout_event = FlowTimeoutPacketEvent(self.flow, curr_packet)
+            main_event_loop.schedule_event_with_delay(timeout_event,
+                                                      FLOW_TIMEOUT_SEC)
+
```
nit: keep two spaces between classes and between functions outside of classes. PyCharm will underline PEP8 issues like this in grey by default (I think the "class FlowTimeoutPacketEvent" would be underlined in this case).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -4,9 +4,13 @@
 """
 
 from abc import ABCMeta, abstractmethod
```
Add some brief comments at the beginning of each class that describe what the Event does.

sharonyang replied:
Unintentionally put it in the initializer and should be fixed now. Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # If the particular packet is currently in transit, then we don't
```
If the packet is in transit, then the fact that the timeout occurred is a sign that it might be lost. So, the case where we immediately return is if the packet is not in transit (since then the packet must have been received).

sharonyang replied:
I think the code was correct. The comment was wrong. Fixed.
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
```
nit: just scheduling a FlowSendPacketsEvent for that timed-out packet, no need to mention what that does internally.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # need to do anything here.
+        if self.packet.packet_id not in self.flow.packets_in_transit:
+            return
+        flow_send_event = FlowSendPacketsEvent(self.flow, [self.packets])
```
List should be "[self.packet]"

lakshbhasin replied:
Before generating the FlowSendPacketsEvent, call flow.handle_packet_loss().

(possible side effect is that this will decrease the window size and so the sanity check in the FlowSendPacketsEvent will fail. Need to figure this out.)

sharonyang replied:
Done.
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # we have finished TCP algorithm setup.
+        # 1. If there is a gap in the list of sent packets by ACK, then that
+        # means that a packet was lost (since we are using selective repeat).
+        # Update window_size accordingly based on TCP algorithm.
```
Updating the window size should just be a call to flow.handle_packet_loss() with the packet that was skipped in the ACK.

sharonyang replied:
Since we have not implemented a TCP algorithm set up here, I'll leave this as a comment.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # 1. If there is a gap in the list of sent packets by ACK, then that
+        # means that a packet was lost (since we are using selective repeat).
+        # Update window_size accordingly based on TCP algorithm.
+        # 2. If there is a duplicate ACK, update window_size and other params
```
Duplicate ACKs are the same as what I mentioned in the note above. Just need to detect if there is a gap, and call handle_packet_loss() on the flow. The flow will be tracking state (depending on TCP algorithm) and will figure out what to do.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # Remove packet in transit.
+        self.flow.packets_in_transit.remove(self.packet.packet_id)
+
+        # TODO(sharon): Base on following description, update the code when
```
The first thing I would do is flow.handle_packet_success() on the packet that was ACK'd. This will at least set in motion the initial window size update based on successful transmission (exact specifics depend on TCP algo).

If packet loss occurs (cases below), then handle_packet_loss(). This is a less common case so should happen after.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # Update window_size accordingly based on TCP algorithm.
+        # 2. If there is a duplicate ACK, update window_size and other params
+        # based on TCP algorithm.
+        # 3. If there is no gap, then there was no packet loss. Update
```
Again, case 3 should happen earlier (as mentioned above).

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        # 2. If there is a duplicate ACK, update window_size and other params
+        # based on TCP algorithm.
+        # 3. If there is no gap, then there was no packet loss. Update
+        # window_size based on TCP algorithm.
```
Add a boolean ivar like had_packet_loss. If a packet was lost, this is set to True, and then flow.handle_packet_loss() will take care of scheduling Events. It's a bit messy but I'm not sure if there's a better way.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # gap = a packet loss (gap in ACK or dupACK)
```
flow.handle_packet_loss() should deal with packet losses and regenerating FlowSendPacketEvents. This is probably cleaner since conditions for retransmit can get complicated.

So you should only do something in schedule_new_events() if had_packet_loss is False. Otherwise, return.

(I think that makes sense and will work.)

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        if not new_packet:
+            # Find the highest ID packet in packets_in_transit and go one more
+            # than that.
+            new_packet_id = max(self.flow.packets_in_transit) + 1
```
This won't work if you removed the old packet ID from self.flow.packets_in_transit. So maybe find the new_packet_id before removing it in run(), and have it as an ivar.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+            new_packet_id = max(self.flow.packets_in_transit) + 1
+            if self.flow.packet_id_exceeds_data(new_packet_id):
+                return
+            new_packet = DataPacket(new_packet_id + 1, self.flow.flow_id,
```
I don't think you need to +1 again.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+                                    self.flow.source_addr, self.flow.dest_addr,
+                                    main_event_loop.global_clock_sec)
+
+        # Schedule a FlowSendPacketsEvent with gap or new max packet.
```
Gaps will be dealt with by flow.handle_packet_loss(). This call here should only be made if there are no packet losses.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/packet.py:
```diff
@@ -53,6 +53,7 @@ def __init__(self, packet_id=None, flow_id=None, source_id=None,
                         start_time_sec=start_time_sec,
                         size_bits=DATA_PACKET_SIZE_BITS)
 
+    # TODO(david): Need __eq__ here to do "packet in list()" in flow.py.
```
You can implement this yourself. Just have a function in Packet that returns False with RoutingPackets (should never have to check these for equality afaik), and for other Packets checks type equality and ID equality.

sharonyang replied:
Done.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        flow_send_event = FlowSendPacketsEvent(self.flow,
```
No need to send anything if packets_to_send is empty, so maybe check that in case. Also, in that case, we would generate a FlowCompleteEvent immediately. See below on what this Event should do.

Edit: Cody said this would be a good idea to add. Or schedule routing table update events in the main event loop itself.

sharonyang replied:
Done.

lakshbhasin replied:
fyi I removed the FlowCompleteEvent stuff here because that should just be triggered when the ACK is received.

lakshbhasin commented on network_simulator/flow.py:
```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+            # than that.
+            new_packet_id = max(self.flow.packets_in_transit) + 1
+            if self.flow.packet_id_exceeds_data(new_packet_id):
+                return
```
If the packet ID exceeds data size, then maybe generate a FlowCompleteEvent. This Event won't do anything in run() or schedule_new_events(), but it'll let the main event loop know when a Flow is done.

sharonyang replied:
Done.

sharonyang committed:
Modified flow event subclasses per Laksh's comments 1 on PR#10.

Commit:
Flow refactoring and other fixes.

Added Packet ID buffer (in case a lost packet cannot immediately be
retransmitted) and cleaned up the code by adding functions in Flow. Also
fixed possibly nasty race conditions (packets_in_transit includes
scheduled DataPackets' IDs too now). Also added preliminary ACK-gap
logic and fixed errors / out-of-date TODOs.

In MainEventLoop, I have also added checks for FlowCompleteEvent.

One thing left to do is to make a PeriodicFlowInterrupt Event for
algorithms like TCP FAST, which update the window size (and schedule new
Packets to send) periodically -- *not* in response to ACKs/losses.
